### PR TITLE
feat(hashicorp-kv): add feature to extend key management service at runtime

### PR DESCRIPTION
**Pull request by NishantJoshi00**
## Description

This PR adds 2 features
- Allowing us to switch between KMS and noKMS at runtime
- Adding a new Key management service to support at runtime (HashiCorp Vault KV2)

## Testing
KMS testing
kms test
HashiCorp testing (kv2 engine)
hc test

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -7,9 +7,10 @@ default-run = "locker"
 # See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
 
 [features]
-default = []
```
Comment (jagan-jaya) [inline Cargo.toml:10]: undo this change

File: src/config.rs [hunk 2]
```diff
@@ -2,20 +2,35 @@ use super::logger::config::Log;
...
 
 #[derive(Clone, serde::Deserialize, Debug)]
 pub struct Config {
     pub server: Server,
     pub database: Database,
     pub secrets: Secrets,
-    #[cfg(feature = "kms")]
-    pub kms: kms::KmsConfig,
+    // #[cfg(feature = "kms")]
+    // pub kms: kms::KmsConfig,
```
Comment (jagan-jaya) [inline src/config.rs:16]:
```suggestion
```

File: src/crypto/hollow.rs [hunk 3]
```diff
@@ -0,0 +1,8 @@
+pub struct Empty;
```
Comment (jagan-jaya) [inline src/crypto/hollow.rs:1]: Rename it to NoEncrytion

File: src/crypto/hollow.rs [hunk 4]
```diff
@@ -0,0 +1,8 @@
...
+
+impl<I: super::FromEncoded> super::Decode<I, String> for Empty {
```
Comment (jagan-jaya) [inline src/crypto/hollow.rs:3]:
```suggestion
impl<I: super::FromEncoded> super::Decode<I, String> for NoEncrytion {
```

File: src/crypto/multiple.rs [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+use futures_util::Future;
+
+use crate::error::{self, KmsError};
+
+pub enum Multiple {
+    #[cfg(feature = "kms")]
+    AwsKms(super::kms::KmsClient),
+    #[cfg(feature = "hashicorp-vault")]
+    VaultKv2(super::vault::HashiCorpVault<super::vault::Kv2>),
+    None(super::hollow::Empty),
```
Comment (jagan-jaya) [inline src/crypto/multiple.rs:12]:
```suggestion
    None(super::hollow::NoEncryption),
```

File: src/crypto/vault.rs
Comment (jagan-jaya) [inline src/crypto/vault.rs:1]: can we rename it to hcvault.rs?

File: src/crypto/vault.rs [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+
+use error_stack::ResultExt;
+use futures_util::Future;
+use vaultrs::client::{VaultClient, VaultClientSettingsBuilder};
+
+use crate::error::{ConfigurationError, KmsError};
+
+#[derive(Clone, Debug, Default, serde::Deserialize)]
+#[serde(default)]
+pub struct VaultConfig {
```
Comment (jagan-jaya) [inline src/crypto/vault.rs:11]:
```suggestion
pub struct HCVaultConfig {
```

File: Cargo.toml [hunk 7]
```diff
@@ -7,9 +7,10 @@ default-run = "locker"
...
+default = ["kms", "hashicorp-vault"]
 release = ["kms", "middleware", "key_custodian", "limit"]
```
Comment (jagan-jaya) [inline Cargo.toml:11]: add `hashicorp-vault` here

**Review by jagan-jaya — APPROVED** [review]

**Review by HyperSwitchers — APPROVED** [review]

File: config/development.toml [hunk 8]
```diff
@@ -63,6 +63,6 @@ Sau7H1Bhzy5G7rwt05LNpU6nFcAGVaZtzl4/+FYfYIulubYjuSEh72yuBHHyvi1/
 -----END RSA PRIVATE KEY-----
 """
 
-[kms]
-region = "us-west-2"
-key_id = "abc"
+# [kms]
```
Comment (jagan-jaya) [inline config/development.toml:66]: mention it in examples.toml

**Review by jagan-jaya — DISMISSED** [review]

**Review by Chethan-rao — DISMISSED** [review]
Aws Kms related files and components could be renamed to contain `Aws` so that its distinguishable from other kms services if added in future. can take up in separate PR

**Review by jagan-jaya — APPROVED** [review]

**Review by Chethan-rao — DISMISSED** [review]

File: src/crypto.rs [hunk 9]
```diff
@@ -16,15 +16,69 @@ pub trait Encode<I, O> {
...
+}
+
+impl FromEncoded for Vec<u8> {
+    fn from_encoded(input: String) -> Option<Self> {
+        hex::decode(input).ok()
+    }
+}
+
 pub mod aes;
+#[cfg(feature = "aws-kms")]
```
Comment (knutties) [inline src/crypto.rs:43]: It might be a good idea to name these features using a common-prefix i.e.

`kms-aws`
`kms-hashicorp-vault` 

to give an indication that they are related.

File: config.example.toml [hunk 10]
```diff
@@ -25,6 +25,21 @@ master_key = "feffe9928665731c6d6a8f9467308308feffe9928665731c6d6a8f9467308308"
...
+# - secrets.master_key
+# - secrets.tenant_public_key
+# - secrets.locker_private_key
+# 
+# Following possible encryption schemes are used, of of them are mutually exclusive, the sections are:
+# - aws_kms (AWS KMS Symmetric Encryption)
+# - vault_kv2 (HashiCorp Vault Secrets Engine KV version 2)
+
+
+[aws_kms]
```
Comment (knutties) [inline config.example.toml:39]: Same comment as the feature flag.  We can name these similar to the feature flag names with `kms-` or `kms_` being the shared common prefix.  Kind of indicates that these are all alternatives.

**Review by jagan-jaya — APPROVED** [review]

**Review by Chethan-rao — APPROVED** [review]
